### PR TITLE
Rework Subxt API to support offline and dynamic transactions 

### LINKED ISSUE #406: Add more dynamic decoding support to `subxt`
Currently `subxt` is optimised towards the compile-time oriented codegen usage; we generate an interface and then use that interface to query compatible nodes. This has many advantages (you benefit from compile-time errors, IDE support, performance and more). The main downside is that the compile-time generated interface is unable to work across different, differing nodes/chains in a general way (parts of it may continue to work, and others will not).

Having the ability to dynamically query nodes across different chains can be very useful in some cases, for example exploratory tooling which wants to interrogate the available interfaces, or tooling akin to polkagot.js which needs to be able to provide a dynamic interface to interact with arbitrary substrate based chains.

These two directions (compiletime, static codegen and dynamic, runtime interaction) are not mutually exclusive either. They would both benefit from the same client (JSONRPC transport/API) code for instance. Another overlap is that `subxt` is used to provide a static interface to a single pallet for the `cargo-contracts` tool, which means being able to dynamically decode `Events` emitted from other pallets that it does not know about (rather than erroring because it's unable to statically decode them).

So, my proposal is that we work towards something like the following:

- Three user facing library crates (names just placeholders); `subxt-static` (which itself uses `subxt-macro` and `subxt-codegen`), `subxt-dynamic` and `subxt-client`. 
    - `subxt-static` is just `subxt` as it stands (with a little runtime decoding moved out of it). 
    - `subxt-dynamic` provides similar functionality to the above, but at runtime. It contains the logic needed to dynamically encode and decode things.
    - `subxt-client` splits out the JSONRPC client API as it currently stands, so that it can be shared between the above.
    - (I wonder whether we would want to share the metadata wrapping in eg `subxt-metadata`, too; perhaps not initially)
- These crates should interop nicely; we can move back and forth between them where it makes sense. As an example, when we fetch the SCALE encoded Event blob, the `cargo-contracts` use case requires that we can dynamically decode (and throw away) any events we don't care about until we get to the one we're looking for (with eg `find_event::<MyEvent>()`).

### LINKED ISSUE #581: Add API to submit transactions using runtime `scale_value::Value`'s
A transaction can basically be represented as a pallet name, call name, and a vec of `scale_value::Value`'s representing each argument to the call. (In theory the whole thing could be represented by a single `Value`, but is there an advantage to that or does it just leave more room for errors?).

Let's add an API to allow transactions of this form to be created and submitted, ideally reusing as much as possible of the existing logic.

### LINKED ISSUE #582: Does it make sense to support encoding/decoding storage lookups as `scale_value::Value`s?
Let's see whether this makes sense to add as part of supporting more dynamic querying.

Propose an API here and then if we agree that it makes sense, we can go ahead and implement it.

### LINKED ISSUE #592: possibility to disable `metadata verification`
Hey,

I have dedicated use-case that I care that some certain calls in a pallet works and optionally that decoding some events.

When an extrinsic is submitted then `subxt` reports `IncompatibleMetadata` but the extrinsic itself was successfully transmitted to the chain.

Thus, I would want to disable this check metadata check and only fail if I get `transaction verification error` when the RPC call fails when using [SubmittableExtrinsic::sign_and_submit_then_watch](https://docs.rs/subxt/latest/subxt/struct.SubmittableExtrinsic.html#method.sign_and_submit_then_watch).

What would be the downside disabling the metadata check completely?  Some events could not be decoded?

### PR DESCRIPTION
**Pull request by jsdw**
- Separate client logic from transaction building logic.
- Support an online and offline client; any functionality which doesn't require an online client can then be used with an offline one instead, for purely offline workflows.
- Once separated, it adds basic support for creating dynamic transactions and such where appropriate ~~(this might land in this PR in an early stage to be built on further)~~.

# Tips for Reviewers

The best place to start is probably by looking at the examples and tests to see what using the new APIs looks like. Inparticular, a new `dynamic_queries.rs` example mirrors some of the other examples but shows how you'd use purely runtime Values and no codegen to do things. 

Next, the `client` folder is the API entry point, and from there you can look at the individual TxClient, StorageClient, ConstantsClient and EventClient to see what methods are available for each and what data is accepted.

These clients generally accept Address/Payload traits, for which we have an implementation used by the static codegen and another used for submitting dynamic Values. This means that the actual API is tied to neither approach.

From here, the constants, events. storage and tx folders contain the respective logic for each area of the API. `storage/storage_address.rs`, `ts/tx_payload.rs`, `constants/constant_address.rs` have the traits and static/dynamic implementations for submitting things to the API.

A `dynamic.rs` file/module exists as a shortcut to all things dynamic (`dynamic::tx()`, `dynamic::storage()`) etc. These are just aliases to the relevant `dynamic` function in each location, so that it's easy to find everything in one place.

This closes #581, closes #406, closes #592 and closes #582

### CODE REVIEW

**Comment by gregdhill** [comment]
@jsdw I'm intrigued by the prospect of dynamic transactions, do you already know what this API will look like? Will it be possible to use subxt without static metadata?

**Comment by jsdw** [comment]
@gregdhill roughly; the core APIs are pretty much finished now in this PR (have a look at something like https://github.com/paritytech/subxt/blob/jsdw-wip-api-changes/examples/examples/balance_transfer.rs; all of the examples are uptodate with the new structure now so they should give a good idea for how things will look).

The overarching idea is to separate building transactions/storage-lookups/constant-lookups from any static codegen. Eg for transactions, now you will build a transaction (via static codegen for instance) and then hand it to the client to submit it. Anything implementing a particular trait can be submitted, and so it'll be quite easy to add support for dynamic transactions too. (and given that it's a trait, you'll be able to use your own data formats to describe transactions if you prefer).

So yup, the goal is that you'll be able to use subxt entirely without any static codegen if you like (but the static codegen is just there to make building transactions/lookups easier/more robust if you want it).

**Comment by jsdw** [comment]
As a side note, this will be a significant change to the APIs (though hopefully one that can be fairly mechanically applied), but it's necessary to do this in order to separate the API client from the static codegen, and I think I've generally been able to simplify quite a lot of things here, and so all in all I hope it'll be a solid foundation to add to which shouldn't need to change much more after this.

**Comment by jsdw** [comment]
@gregdhill I've gotten to the point of having an example showing some dynamic queries now; see:

https://github.com/paritytech/subxt/blob/jsdw-wip-api-changes/examples/examples/dynamic_queries.rs

**Review by lexnv — APPROVED** [review]
👍 LGTM!

File: testing/integration-tests/src/events/mod.rs [hunk 1]
```diff
@@ -4,47 +4,49 @@
...
     test_context,
+    utils::wait_for_blocks,
 };
 use futures::StreamExt;
 use sp_keyring::AccountKeyring;
 
 // Check that we can subscribe to non-finalized block events.
 #[tokio::test]
-async fn non_finalized_block_subscription() -> Result<(), subxt::BasicError> {
-    tracing_subscriber::fmt::try_init().ok();
```
[the remark this answers is not in the document]
Comment (jsdw) [reply testing/integration-tests/src/events/mod.rs:19]: Yeah; I got rid of them because the test output became unnecessarily verbose (but left the crate so that we could add it back on a case-by-case basis for debugging tests)

File: subxt/src/error.rs [hunk 2]
```diff
@@ -2,27 +2,28 @@
...
+pub use sp_core::crypto::SecretStringError;
+pub use sp_runtime::transaction_validity::TransactionValidityError;
 
 /// The underlying error enum, generic over the type held by the `Runtime`
-/// variant. Prefer to use the [`Error<E>`] and [`BasicError`] aliases over
+/// variant. Prefer to use the [`Error<E>`] and [`Error`] aliases over
 /// using this type directly.
 #[derive(Debug, thiserror::Error)]
-pub enum GenericError<E> {
+pub enum Error {
```
[the remark this answers is not in the document]
Comment (jsdw) [reply subxt/src/error.rs:26]: I was happy to get rid of some generic params in this!

File: subxt/src/events/events_client.rs [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+
+impl<T, Client> EventsClient<T, Client>
+where
+    T: Config,
+    Client: OnlineClientT<T>,
+{
+    /// Obtain events at some block hash.
+    pub fn at(
+        &self,
+        block_hash: T::Hash,
```
[the remark this answers is not in the document]
Comment (jsdw) [reply subxt/src/events/events_client.rs:56]: Good idea; that makes sense to me!

File: subxt/src/events/events_client.rs [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+where
+    T: Config,
+    Client: OnlineClientT<T>,
+{
+    let event_bytes = client
+        .rpc()
+        .storage(&*system_events_key().0, Some(block_hash))
+        .await?
+        .map(|e| e.0)
+        .unwrap_or_else(Vec::new);
```
[the remark this answers is not in the document]
Comment (jsdw) [reply subxt/src/events/events_client.rs:132]: Yeah, I'd pretty much always expect there to be events (but I won't hold it against a node if it doesn't produce anything either; there is no reason why there must always be events in a block really!). But yeah, I wouldn't see a lack of events as being something that's worth optimising for really :)

File: subxt/src/events/events_client.rs [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+        sub,
+    );
+
+    Ok(EventSubscription::new(client, Box::pin(block_subscription)))
+}
+
+/// Note: This is exposed for testing but is not considered stable and may change
+/// without notice in a patch release.
+#[doc(hidden)]
+pub fn subscribe_to_block_headers_filling_in_gaps<T, Client, S, E>(
```
[the remark this answers is not in the document]
Comment (jsdw) [reply subxt/src/events/events_client.rs:180]: Nar; this is used above in "real" code too, but maybe it doesn't need to be public any more; def worth checking whether it no longer needs to be exposed!

File: subxt/src/events/events_type.rs [hunk 6]
```diff
@@ -206,138 +119,163 @@ impl<'a, T: Config, Evs: Decode> Events<T, Evs> {
...
+        self.phase
+    }
+
+    /// What index is this event in the stored events for this block.
+    pub fn index(&self) -> u32 {
+        self.index
+    }
+
+    /// The index of the pallet that the event originated from.
+    pub fn pallet_index(&self) -> u8 {
```
[the remark this answers is not in the document]
Comment (jsdw) [reply subxt/src/events/events_type.rs:198]: Actually, they should never panic (see line 302: `let mut event_bytes = vec![pallet_index, variant_index];`); we'll catch such errors when decoding into this thing. But I should add a comment or something here to make that clear, since it's a legitimate concern looking at the code for sure!

File: subxt/src/tx/tx_progress.rs [hunk 7]
```diff
@@ -502,15 +447,123 @@ impl<T: Config, Evs: Decode> TransactionEvents<T, Evs> {
...
+                    return DispatchError::Other(bytes.to_vec())
+                }
+            };
+            CurrentModuleError {
+                index: old_e.index,
+                error: [old_e.error, 0, 0, 0],
+            }
+        }
+    };
+    println!(
```
[the remark this answers is not in the document]
Comment (jsdw) [reply subxt/src/tx/tx_progress.rs:548]: Ooh good catch; this should be deleted! accidental debugging stuff :)

File: testing/integration-tests/src/utils/context.rs [hunk 8]
```diff
@@ -32,35 +26,20 @@ pub async fn test_node_process_with(
...
-    }
-}
+pub type TestContext = TestNodeProcess<SubstrateConfig>;
 
 pub async fn test_context() -> TestContext {
-    tracing_subscriber::fmt::try_init().ok();
-    let node_proc = test_node_process_with(AccountKeyring::Alice).await;
-    let api = node_proc.client().clone().to_runtime_api();
-    TestContext { node_proc, api }
+    // tracing_subscriber::fmt::try_init().ok();
```
[the remark this answers is not in the document]
Comment (jsdw) [reply testing/integration-tests/src/utils/context.rs:39]: Yup :)

File: subxt/src/events/events_client.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jsdw) [reply subxt/src/events/events_client.rs:180]: I can't see a nice way to make it only `pub` for tests without duplicating the entire signature, so I'll leave it be unless you have any other ideas :)

File: subxt/src/events/events_client.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jsdw) [reply subxt/src/events/events_client.rs:56]: The `Events` object expects a hash and not an option, so I added a little code to get the latest hash manually if `None` is provided. Having an option still makes sense to me though; it's consistent with all of the RPC methods!

**Comment by jsdw** [comment]
When this is merged, notw that #607 updates how to decode and present Events to make use of alloc free decoding.

File: examples/examples/custom_type_derives.rs [hunk 9]
```diff
@@ -2,8 +2,7 @@
 // This file is dual-licensed as Apache-2.0 or GPL-3.0.
 // see LICENSE for license details.
 
-//! Example verified against polkadot 0.9.18-4542a603cc-aarch64-macos.
-
+//! Example verified against polkadot polkadot 0.9.25-5174e9ae75b.
```
Comment (niklasad1) [inline examples/examples/custom_type_derives.rs:5]: I reckon that it would be useful to add an example for `overriding types` as well not really related to this PR but could be part of this example perhaps....

File: examples/examples/balance_transfer.rs [hunk 10]
```diff
@@ -26,23 +25,20 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
-        .to_runtime_api::<polkadot::RuntimeApi<DefaultConfig, PolkadotExtrinsicParams<DefaultConfig>>>();
-
-    // Submit the `transfer` extrinsic from Alice's account to Bob's.
     let dest = AccountKeyring::Bob.to_account_id().into();
 
-    // Obtain an extrinsic, calling the "transfer" function in
-    // the "balances" pallet.
-    let extrinsic = api.tx().balances().transfer(dest, 123_456_789_012_345)?;
+    // Create a client to use:
+    let api = OnlineClient::<PolkadotConfig>::new().await?;
```
Comment (niklasad1) [inline examples/examples/balance_transfer.rs:31]: so one doesn't have to call `to_runtime_api()` anymore nice

File: examples/examples/balance_transfer.rs [hunk 11]
```diff
@@ -2,20 +2,19 @@
 // This file is dual-licensed as Apache-2.0 or GPL-3.0.
 // see LICENSE for license details.
 
-//! To run this example, a local polkadot node should be running. Example verified against polkadot 0.9.18-4542a603cc-aarch64-macos.
+//! To run this example, a local polkadot node should be running. Example verified against polkadot polkadot 0.9.25-5174e9ae75b.
```
Comment (niklasad1) [inline examples/examples/balance_transfer.rs:5]: one additional with example `OfflineClient` would be great

File: codegen/src/api/storage.rs [hunk 12]
```diff
@@ -255,133 +169,85 @@ fn generate_storage_entry_fns(
...
-    let storage_entry_type = quote! {
-        #entry_struct
-        impl ::subxt::StorageEntry for #entry_struct_ident #anon_lifetime {
-            #storage_entry_impl
-        }
+    let key_args = fields.iter().map(|(field_name, field_type)| {
+        // The field type is translated from `std::vec::Vec<T>` to `[T]`. We apply
+        // AsRef to all types, so this just makes it a little more ergonomic.
+        //
+        // TODO [jsdw]: Support mappings like `String -> str` too for better Borrow
```
Comment (niklasad1) [inline codegen/src/api/storage.rs:185]: file an issue?

Comment (niklasad1) [inline codegen/src/api/storage.rs:185]:
```suggestion
        // TODO [jsdw]: Support mappings like `String -> str` too for better borrow
```

File: subxt/src/client/online_client.rs [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+use std::sync::Arc;
+
+/// A trait representing a client that can perform
+/// online actions.
+pub trait OnlineClientT<T: Config>: OfflineClientT<T> {
+    /// Return an RPC client that can be used to communicate with a node.
+    fn rpc(&self) -> &Rpc<T>;
+}
+
+/// A client capable of performing offline or online operations.
```
Comment (niklasad1) [inline subxt/src/client/online_client.rs:35]: I find these docs a bit confusing i.e, `online or offline`.

I suggest that some like: `A client capable of performing online operations but it can be converted to an offline client`

File: subxt/src/client/online_client.rs [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+impl<T: Config> std::fmt::Debug for OnlineClient<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Client")
+            .field("rpc", &"<Rpc>")
+            .field("inner", &self.inner)
+            .finish()
+    }
+}
+
+impl<T: Config> OnlineClient<T> {
```
Comment (niklasad1) [inline subxt/src/client/online_client.rs:60]: I really like these "new/changed" APIs

File: subxt/src/config.rs [hunk 15]
```diff
@@ -97,4 +105,46 @@ impl Config for DefaultConfig {
         sp_runtime::generic::Header<Self::BlockNumber, sp_runtime::traits::BlakeTwo256>;
     type Signature = sp_runtime::MultiSignature;
     type Extrinsic = sp_runtime::OpaqueExtrinsic;
+    type ExtrinsicParams = crate::tx::SubstrateExtrinsicParams<Self>;
+}
+
+/// Default set of commonly used types by Polkadot nodes.
+pub type PolkadotConfig = WithExtrinsicParams<
+    SubstrateConfig,
```
Comment (niklasad1) [inline subxt/src/config.rs:113]: I don't know how that works but for polkadot the address type is different https://github.com/paritytech/polkadot/blob/master/runtime/polkadot/src/lib.rs#L1488

File: subxt/src/constants/constant_address.rs [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+    fn pallet_name(&self) -> &str;
+
+    /// The name of the constant in a given pallet.
+    fn constant_name(&self) -> &str;
+
+    /// An optional hash which, if present, will be checked against
+    /// the node metadata to confirm that the return type matches what
+    /// we are expecting.
+    fn validation_hash(&self) -> Option<[u8; 32]> {
+        None
```
Comment (niklasad1) [inline subxt/src/constants/constant_address.rs:27]: why default impl? 

I have been bitten by that a couple of times and prefer explict impls for each type :P

File: subxt/src/constants/constants_client.rs [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+        &self,
+        address: &Address,
+    ) -> Result<(), Error> {
+        if let Some(actual_hash) = address.validation_hash() {
+            let expected_hash = self
+                .client
+                .metadata()
+                .constant_hash(address.pallet_name(), address.constant_name())?;
+            if actual_hash != expected_hash {
+                return Err(MetadataError::IncompatibleMetadata.into())
```
Comment (niklasad1) [inline subxt/src/constants/constants_client.rs:50]: When this fails it would be super useful if the actual pallet constant name is embedded in the error.

This is hard to debug otherwise.

File: subxt/src/events/event_subscription.rs [hunk 18]
```diff
@@ -144,52 +44,80 @@ pub type EventSub<Item> = Subscription<Item>;
...
 
     /// Return only specific events matching the tuple of 1 or more event
     /// types that has been provided as the `Filter` type parameter.
-    pub fn filter_events<Filter: EventFilter>(self) -> FilterEvents<'a, Self, T, Filter> {
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// use futures::StreamExt;
+    /// use subxt::{ OnlineClient, PolkadotConfig };
```
Comment (niklasad1) [inline subxt/src/events/event_subscription.rs:78]:
```suggestion
    /// use subxt::{OnlineClient, PolkadotConfig};
```

File: subxt/src/events/event_subscription.rs [hunk 19]
```diff
@@ -30,112 +26,16 @@ use std::{
...
-        // Return a combination of any previous headers plus the new header.
-        Either::Right(previous_headers.chain(stream::once(async { Ok(header) })))
-    })
-}
-
 /// A `jsonrpsee` Subscription. This forms a part of the `EventSubscription` type handed back
 /// in codegen from `subscribe_finalized`, and is exposed to be used in codegen.
 #[doc(hidden)]
-pub type FinalizedEventSub<'a, Header> = BoxStream<'a, Result<Header, BasicError>>;
+pub type FinalizedEventSub<Header> = BoxStream<'static, Result<Header, Error>>;
```
Comment (niklasad1) [inline subxt/src/events/event_subscription.rs:38]: what's the reason of the change? 
ease of use with less lifetime stuff?

File: subxt/src/metadata/metadata_type.rs [hunk 20]
```diff
@@ -273,9 +294,10 @@ impl PalletMetadata {
 /// Metadata for specific events.
 #[derive(Clone, Debug)]
 pub struct EventMetadata {
-    pallet: String,
+    pallet: Arc<str>,
```
Comment (niklasad1) [inline subxt/src/metadata/metadata_type.rs:297]: can you explain why this is changed to an `Arc`?

File: examples/examples/balance_transfer.rs (hunk 11 above)
Comment (jsdw) [reply examples/examples/balance_transfer.rs:5]: What sort of thing did you have in mind? Maybe an example where we create a signed extrinsic offline and then just print out the hex for it?

File: subxt/src/events/events_client.rs [hunk 21]
```diff
@@ -0,0 +1,249 @@
...
+impl<T, Client> EventsClient<T, Client>
+where
+    T: Config,
+    Client: OnlineClientT<T>,
+{
+    /// Obtain events at some block hash.
+    pub fn at(
+        &self,
+        block_hash: Option<T::Hash>,
+    ) -> impl Future<Output = Result<Events<T>, Error>> + Send + 'static {
```
[the remark this answers is not in the document]
Comment (jsdw) [reply subxt/src/events/events_client.rs:57]: Just because anm `async fn` would have an implicit lifetime tied to self; here we clone the client out of self so that the returned future can have a `'static` lifetime. I think the motivating example here was so you could do stuff like `let fut = api.events().at(...)` and then `futures::join!(fut, fut2..)`; with a lifetime tied to self I think it would complain that `api.events()` was temporary and didn't live long enough if self was needlessly borrowed from it

File: subxt/src/metadata/metadata_type.rs (hunk 20 above)
Comment (jsdw) [reply subxt/src/metadata/metadata_type.rs:297]: Ah sure; Basically because the `pallet` name is duplicated across every event in a pallet ,so instead of `clone()`int a String to store with every event I just stuck it behind an arc so that it's an increment instead.

Probably worth a comment in the code :)

File: examples/examples/balance_transfer.rs (hunk 11 above)
Comment (niklasad1) [reply examples/examples/balance_transfer.rs:5]: the simplest possible how to instantiate it and use it just to demonstrate how the API works... because I don't understand how it will work i.e, how would me as user use the "offline extrinsic" display QR code, send via USB/UART, etc?

create an signed extrinsic sounds good to me....

File: subxt/src/config.rs (hunk 15 above)
Comment (jsdw) [reply subxt/src/config.rs:113]: Ooohhh is it? 

It looks like both are `MultiAddress<AccountId32, _>`'s at least, but that second generic looks like a u32 in substrate and () in polkadot? I guess this means that the byte encoding is identical but polkadot doesn't support Index addresses.

Maybe we should change the PolkadotConfig to match this then, just to make sure people use the right types?

File: subxt/src/events/event_subscription.rs (hunk 19 above)
Comment (jsdw) [reply subxt/src/events/event_subscription.rs:38]: Yeah, I don't remember exactly but think I just wanted to simplify the lifetime stuff, and doing the odd `Arc::clone` instead made it easier to pass the client around and not have any lifetime issues moving stuff between threads and whatnot.

Comment (jsdw) [reply subxt/src/events/event_subscription.rs:38]: (oh and to add; with that in place I guess this no longer needed to be tied to the lifetime of the client, since the client is no longer a reference)

File: subxt/src/constants/constants_client.rs (hunk 17 above)
Comment (jsdw) [reply subxt/src/constants/constants_client.rs:50]: Yeahh Andrew raised this too; it's a separate issue somewhere so I didn't want to think about it in this PR!

File: subxt/src/constants/constant_address.rs (hunk 16 above)
Comment (jsdw) [reply subxt/src/constants/constant_address.rs:27]: This is just because we only really ever expect to use `validation_hash` internally to allow us to check our static codegen against our metadata hashes. By default, anybody who implements this trait themselves will not be doing their own static codegen and will instead be using their own custom types which won't have a sensible impl for this :)

The worst case of not implementing it is that we don't do our extra validation check anyway

File: examples/examples/custom_type_derives.rs (hunk 9 above)
Comment (jsdw) [reply examples/examples/custom_type_derives.rs:5]: Let's add an issue to do this :)

File: examples/examples/balance_transfer.rs (hunk 11 above)
Comment (jsdw) [reply examples/examples/balance_transfer.rs:5]: Sounds good to me!

I think let's create an issue for this to avoid doing more work on this PR :)

File: subxt/src/config.rs (hunk 15 above)
Comment (niklasad1) [reply subxt/src/config.rs:113]: yeah, I meant that that second generic param is different as you said.

doesn't seem to have any implications when I have used subxt but might be better match the one in polkadot ^^

File: examples/examples/balance_transfer.rs (hunk 11 above)
Comment (jsdw) [reply examples/examples/balance_transfer.rs:5]: (I'll make one!)

File: subxt/src/client/online_client.rs (hunk 13 above)
Comment (jsdw) [reply subxt/src/client/online_client.rs:35]: I agree that the comment is a bit naff; it was trying to say that this client can be used in any place that requires an `OnlineClientT` or an `OfflineClientT` (and yes, it can also be converted into an `OfflineClient` if you want to limit it to only being accepted in places needing an `OfflineClientT`)

File: subxt/src/config.rs (hunk 15 above)
Comment (jsdw) [reply subxt/src/config.rs:113]: Yeah; I guess it has no impact on the SCALE encoding, so you'd only bump into issues trying to pass eg a `MultiAddress::Index(123)` in to a polkadot node. Otherwise you'll be fine.

I'll add a separate issue to fix it; it's been like this for ages :)

File: subxt/src/client/online_client.rs (hunk 14 above)
Comment (jsdw) [reply subxt/src/client/online_client.rs:60]: Ah I'm glad to hear it!

File: examples/examples/custom_type_derives.rs (hunk 9 above)
Comment (jsdw) [reply examples/examples/custom_type_derives.rs:5]: Actually I'll do it if you haven't already; I have a list of issues to make anyway :)

**Comment by jsdw** [comment]
Ah, the checks ran foul of some recent substrate build change that we picked up on a couple of nights ago; guess I'll have to fix it now!

File: subxt/src/metadata/metadata_type.rs (hunk 20 above)
Comment (jsdw) [reply subxt/src/metadata/metadata_type.rs:299]: (comment added)

File: examples/examples/balance_transfer.rs (hunk 10 above)
Comment (jsdw) [reply examples/examples/balance_transfer.rs:31]: Yup; no more "hard" link between static codegen and API :)

File: subxt/src/storage/utils.rs [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+    metadata: &Metadata,
+) -> Result<Vec<u8>, Error> {
+    let mut bytes = Vec::new();
+    storage_address_root_bytes(addr, &mut bytes);
+    addr.append_entry_bytes(metadata, &mut bytes)?;
+    Ok(bytes)
+}
+
+/// Outputs a vector containing the [`storage_address_root_bytes`] bytes.
+pub fn storage_address_to_root_bytes<Address: StorageAddress>(addr: &Address) -> Vec<u8> {
```
Comment (niklasad1) [inline subxt/src/storage/utils.rs:38]:
```suggestion
pub fn storage_address_root_to_bytes<Address: StorageAddress>(addr: &Address) -> Vec<u8> {
```

File: subxt/src/events/events_client.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (niklasad1) [reply subxt/src/events/events_client.rs:57]: ah, I see

File: subxt/src/storage/utils.rs [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+
+use super::StorageAddress;
+use crate::{
+    error::Error,
+    metadata::Metadata,
+};
+
+/// Return the root of a given [`StorageAddress`]: hash the pallet name and entry name
+/// and append those bytes to the output.
+pub fn storage_address_root_bytes<Address: StorageAddress>(
```
[the remark this answers is not in the document]
Comment (jsdw) [reply subxt/src/storage/utils.rs]: Good catch; these names were rubbish! I changed them both :)

File: subxt/src/client/offline_client.rs [hunk 24]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    /// Return the runtime version.
+    pub fn runtime_version(&self) -> RuntimeVersion {
+        self.inner.runtime_version.clone()
+    }
+
+    /// Return the [`Metadata`] used in this client.
+    pub fn metadata(&self) -> Metadata {
+        self.inner.metadata.clone()
```
Comment (ascjones) [inline subxt/src/client/offline_client.rs:93]: Might be better to return borrowed `&Metadata` since it can be quite big? More of an issue in resource constrained environments where `OfflineClient` is more likely to be used.

User can always manually clone if required.

File: subxt/src/error.rs [hunk 25]
```diff
@@ -49,102 +50,68 @@ pub enum GenericError<E> {
     Metadata(#[from] MetadataError),
     /// Runtime error.
     #[error("Runtime error: {0:?}")]
-    Runtime(E),
-    /// Events decoding error.
-    #[error("Events decoding error: {0}")]
-    EventsDecoding(#[from] DecodeError),
+    Runtime(DispatchError),
```
Comment (ascjones) [inline subxt/src/error.rs:53]: Ah yes much nicer :ok_hand:

File: subxt/src/tx/tx_progress.rs [hunk 26]
```diff
@@ -502,15 +447,120 @@ impl<T: Config, Evs: Decode> TransactionEvents<T, Evs> {
...
+        }
+    };
+
+    let error_details = match metadata.error(err.index, err.error[0]) {
+        Ok(details) => details,
+        Err(_) => {
+            tracing::warn!("Can't decode error: sp_runtime::DispatchError::Module details do not match known information");
+            return DispatchError::Other(bytes.to_vec())
+        }
+    };
```
Comment (ascjones) [inline subxt/src/tx/tx_progress.rs:555]: :heart:

File: testing/integration-tests/src/client/mod.rs [hunk 27]
```diff
@@ -113,9 +121,96 @@ async fn test_iter() {
...
+        .transfer(bob.account_id().clone().into(), 10_000);
+
+    let signed_extrinsic = api
+        .tx()
+        .create_signed(&tx, &alice, Default::default())
+        .await
+        .unwrap();
+
+    api.rpc()
+        .dry_run(signed_extrinsic.encoded(), None)
```
Comment (ascjones) [inline testing/integration-tests/src/client/mod.rs:153]: Awesome!!!!!!!

**Review by shekohex — APPROVED** [review]
Great Job!! Really looking forward to use this release!

File: subxt/src/client/offline_client.rs (hunk 24 above)
Comment (jsdw) [reply subxt/src/client/offline_client.rs:93]: It's all behind an Arc anyway; the reason we return a clone is so that the OnlineClient (which has it behind an RwLock) can also meet this type signature (It couldn't return a reference from behidn the lock; it'd need to return a guard)

File: subxt/src/tx/tx_progress.rs [hunk 28]
```diff
@@ -502,15 +447,120 @@ impl<T: Config, Evs: Decode> TransactionEvents<T, Evs> {
...
-    pub fn has<Ev: crate::Event>(&self) -> Result<bool, BasicError> {
+    pub fn has<Ev: StaticEvent>(&self) -> Result<bool, Error> {
         Ok(self.find::<Ev>().next().transpose()?.is_some())
     }
 }
+
+// Attempt to decode a DispatchError, returning either the ModuleError it decodes to,
+// along with additional details on the error, or returning the raw bytes to work with
+// downstream if we couldn't decode it.
+fn decode_dispatch_error(metadata: &Metadata, bytes: &[u8]) -> DispatchError {
```
Comment (ascjones) [inline subxt/src/tx/tx_progress.rs:466]: I may need access to this, see: https://github.com/paritytech/cargo-contract/blob/master/src/cmd/extrinsics/mod.rs#L246.

Essentially I want to decode the `DispatchError` returned from an RPC call rather than this extrinsic.

Could we hang it off `Metadata`? Or `DispatchError`?

**Review by ascjones — APPROVED** [review]
Bravo! Really like this new API :+1:

File: subxt/src/tx/tx_progress.rs (hunk 28 above)
Comment (jsdw) [reply subxt/src/tx/tx_progress.rs]: Ooh yup, sounds very reasonable to me! I stuck it onto `DispatchError`

File: subxt/src/tx/tx_client.rs [hunk 29]
```diff
@@ -0,0 +1,331 @@
...
+    ///
+    /// # Note
+    ///
+    /// Success does not mean the extrinsic has been included in the block, just that it is valid
+    /// and has been included in the transaction pool.
+    pub async fn sign_and_submit<Call>(
+        &self,
+        call: &Call,
+        signer: &(dyn Signer<T> + Send + Sync),
+        other_params: <T::ExtrinsicParams as ExtrinsicParams<T::Index, T::Hash>>::OtherParams,
```
[the remark this answers is not in the document]
Comment (jsdw) [reply subxt/src/tx/tx_client.rs:267]: I had a go at this, and where I landed was:

- If we can accept a reference to OtherParams when building ExtrinsicParams that'd be nice, and probably reflexct how it's used the best. But it turns out to be a pain to do so the complexity didn't feel worth while.
- If we don't, then we need to do a `.clone()` call internally, and I'd rather push that to the user (and then impls can be copy potentially and it's no biggie, or they aren't and better the user has the chance to pass ownership rather than force a clone.

So for now I'd suggest elaving it as it is, but we might be able to revisit and improve!
